Makefile

```diff
@@ -16,75 +15,141 @@ BRIGADE_GITHUB_GW_EXTERNAL_PORT := 7744
+
+amd64_DOCKER_ARCH = amd64
+arm64_DOCKER_ARCH = arm64v8
+arm32_DOCKER_ARCH = arm32v6
+
+amd64_DOCKER_ANNOTATIONS = --os linux --arch amd64
+arm64_DOCKER_ANNOTATIONS = --os linux --arch arm64 --variant v8
+arm32_DOCKER_ANNOTATIONS = --os linux --arch arm
+
+.PHONY: build build-images push-images dockerfiles
+
+define GO_CMD_TARGETS
+# Build command
+$1/rootfs/$2/$1: vendor $(shell find $1 -type f -name '*.go')
+	GOOS=linux GOARCH=$$($2_GOARCH) CGO_ENABLED=0 go build -ldflags '$(LDFLAGS)' -o ./$1/rootfs/$2/$1 ./$1/cmd/$1
+
+build-$2 $1-$2-image: $1/rootfs/$2/$1
+endef
+
+define DOCKER_IMAGE_TARGETS
+# Build image for a single arch
+.PHONY: $1-images push-$1-images
+$1-$2-image: $1/Dockerfile.$2
+	docker build $(DOCKER_BUILD_FLAGS) -t $(DOCKER_REGISTRY)/$1:$2-$(IMAGE_TAG) -f $1/Dockerfile.$2 $1
+
+# Push image for a single arch
+push-$1-$2-image: $1-$2-image
+	docker push $(DOCKER_REGISTRY)/$1:$2-$(IMAGE_TAG)
+
+# Create arch dockerfile
+$1/Dockerfile.$2: $1/Dockerfile
+	@cat $1/Dockerfile | ARCH="$2" QEMU_IMPORT="$$($2_QEMU_IMPORT)" DOCKER_ARCH="$$($2_DOCKER_ARCH)" envsubst '$$$${ARCH} $$$${QEMU_IMPORT} $$$${DOCKER_ARCH}' > $1/Dockerfile.$2
+
+dockerfiles: $1/Dockerfile.$2
+build-$2-images $1-images: $1-$2-image
+push-$2-images push-$1-images: push-$1-$2-image
+endef
+
+define DOCKER_MANIFEST_TARGETS
+.PHONY: create-$1-manifest $1
+
+$1: $1-images
+
+create-$1-manifest: push-$1-images
+	# Remove old manifest if present
+	@if [ -d $(HOME)/.docker/manifests/docker.io_$(DOCKER_REGISTRY)_$1-$(IMAGE_VERSION) ]; then \
+		rm -rf $(HOME)/.docker/manifests/docker.io_$(DOCKER_REGISTRY)_$1-$(IMAGE_VERSION); \
+	fi
+
+	docker manifest create $(DOCKER_REGISTRY)/$1:$(IMAGE_TAG) $(foreach arch,$(ARCHS),$(DOCKER_REGISTRY)/$1:$(arch)-$(IMAGE_TAG))
+	$(foreach arch,$(ARCHS),docker manifest annotate $$($(arch)_DOCKER_ANNOTATIONS) $(DOCKER_REGISTRY)/$1:$(IMAGE_TAG) $(DOCKER_REGISTRY)/$1:$(arch)-$(IMAGE_TAG);)
+	docker manifest push $(DOCKER_REGISTRY)/$1:$(IMAGE_TAG)
+endef
+
+define BRIG_STATIC_ARCH_TARGETS
+.PHONY: build-$3-static
+build: build-$3-static
+
+ifeq ($1,windows)
+ifeq ($2,amd64)
+build-$3-static: bin/$3-$1-$2.exe
+bin/$3-$1-$2.exe: vendor $(shell find $3 -type f -name '*.go')
+	GOOS=$1 GOARCH=$2 CGO_ENABLED=0 go build -ldflags '$(LDFLAGS)' -o ./$$@ ./$3/cmd/$3
+endif
+endif
+
+ifeq ($1,darwin)
+ifeq ($2,amd64)
+build-$3-static: bin/$3-$1-$2
+bin/$3-$1-$2: vendor $(shell find $3 -type f -name '*.go')
+	GOOS=$1 GOARCH=$2 CGO_ENABLED=0 go build -ldflags '$(LDFLAGS)' -o ./$$@ ./$3/cmd/$3
+endif
+endif
+
+ifeq ($1,linux)
+build-$3-static: bin/$3-$1-$2
+bin/$3-$1-$2: vendor $(shell find $3 -type f -name '*.go')
+	GOOS=$1 GOARCH=$$($2_GOARCH) CGO_ENABLED=0 go build -ldflags '$(LDFLAGS)' -o ./$$@ ./$3/cmd/$3
+endif
+endef
+
+define BRIG_STATIC_TARGETS
+$(foreach cmd,$(BINS),$(eval $(call BRIG_STATIC_ARCH_TARGETS,$2,$1,$(cmd))))
+endef
+
+define GO_ARCH_TARGETS
```

Comment:
I'm wondering if we move lines 128-140 further up at the top, before the definitions of the functions that are invoked in each `foreach` statement (`BRIG_STATIC_TARGETS`, etc.).

Then the flow to understand would be: seeing these meta targets and foreach loops, then digging in to each looped function for how each parameterized target is built...

Reply:
I'm an F#-er, so I'm used to reading files with the exports at the bottom, and the implementation details at the top, hence the ordering (up to a point where I confused myself I think). Will try to move targets that are more likely to be called higher up.